src/common/variant.h

```diff
@@ -0,0 +1,135 @@
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// Variant wrapper class.
+
+#pragma once
+
+//local headers
+
+//third party headers
+#include <boost/blank.hpp>
+#include <boost/none_t.hpp>
+#include <boost/variant/apply_visitor.hpp>
+#include <boost/variant/get.hpp>
+#include <boost/variant/static_visitor.hpp>
+#include <boost/variant/variant.hpp>
+
+//standard headers
+#include <stdexcept>
+#include <type_traits>
+
+//forward declarations
+
+
+namespace tools
+{
+
+////
+// variant: convenience wrapper around boost::variant with a cleaner interface
+// - the value may be assigned to but is otherwise read-only
+// - the variant is 'optional' - an empty variant will evaluate to 'false' and an initialized variant will be 'true'
+///
+template <typename ResultT>
+struct variant_static_visitor : public boost::static_visitor<ResultT>
+{
+    /// provide visitation for empty variants
+    /// - add this to your visitor with: using variant_static_visitor::operator();
+    ResultT operator()(const boost::blank) const
+    {
+        throw std::runtime_error("variant: tried to visit an empty variant.");
+        static const ResultT dummy{};
+        return dummy;
+    }
+};
+
+template <typename... Types>
+class variant final
+{
+    using VType = boost::variant<boost::blank, Types...>;
+
+public:
+//constructors
+    /// default constructor
+    variant() = default;
+    variant(boost::none_t) : variant{} {}  //act like boost::optional
+
+    /// construct from variant type
+    template <typename T>
+    variant(const T &value) : m_value{value} {}
+
+//overloaded operators
+    /// boolean operator: true if the variant isn't empty/uninitialized
+    operator bool() const { return m_value.which() != 0; }
+
+//member functions
+    /// check if empty/uninitialized
+    bool is_empty() const { return !static_cast<bool>(*this); }
+
+    /// check the variant type
+    template <typename T>
+    bool is_type() const { return boost::strict_get<T>(&m_value) != nullptr; }
+
+    /// get a read-only handle to the embedded value
+    template <typename T>
+    const T& unwrap() const
+    {
+        const T *value_ptr{boost::strict_get<T>(&m_value)};
+        if (!value_ptr) throw std::runtime_error("variant: tried to access value of incorrect type.");
+        return *value_ptr;
+    }
+
+    /// get the type index of the current partial signature
+    int type_index() const { return m_value.which(); }
+
+    /// get the type index of a requested type (compile error for invalid types)
+    template <typename T>
+    static int type_index_of()
+    {
+        static const int type_index_of_T{VType{T{}}.which()};  //todo: this is slower than it needs to be
+        return type_index_of_T;
+    }
+
+    /// check if two variants have the same type
+    static bool same_type(const variant<Types...> &v1, const variant<Types...> &v2)
```

Comment:
Static matches the semantics - it's a function called on two variants, not a behavior of one variant.

Reply:
I forgot about ADL briefly, are you invoking this like:
```c++
same_type(val1, val2);
```
? This should work, provided no one else adds a `same_type` function to the namespace. This could be a standalone function - I don't see anything wrong with that either.

Reply:
Nevermind, I would just leave as-is unless it becomes an issue. Might be confusing for the person to change though.

Reply:
[This](https://github.com/UkoeHB/monero/blob/6e7b72ccc3c576f2deff936f1dd4f085a9064798/src/multisig/multisig_signing_helper_utils.cpp#L789) is the only place it's used right now.